File: swift/Sources/CircleAI/Hr.swift
// Hr.swift
//
// Port of the HR vertical from src/CircleAI.HR/HRPrimitives.cs and the static
// domain-context constants from HRDomainContext.cs:
//   • Employee, LeaveRequest, PerformanceReview — domain records
//   • IHRBoard              — hiring, leave management, performance reviews
//   • InMemoryHRBoard       — deterministic in-memory impl
//   • HRDomainContext       — system-prompt snippet + flags
//
// The Companion-facing wrapper (HRCompanionAdapter) is intentionally NOT ported
// (it wraps the Companion session infrastructure, not board state).
//
// Porting notes:
//   • `decimal` → `Decimal`; `DateTime` → `Date`.
//   • `DecideLeave` on an unknown request throws → `HRError.unknownLeaveRequest`.
//   • `Employees` is ordered ascending by Name. `PendingLeaves` filters
//     case-insensitively on Status == "Pending".
//   • `AvgRatingFor` averages RatingOutOf5 for the employee; 0.0 when none.
//   • All state guarded by a single `NSLock`.

import Foundation

// MARK: - Records

/// An employee.
public struct Employee: Sendable, Equatable, Codable {
    public let employeeId: String
    public let name: String
    public let role: String
    public let hiredOn: Date
    public let salary: Decimal
    public let currency: String

    public init(employeeId: String, name: String, role: String, hiredOn: Date, salary: Decimal, currency: String) {
        self.employeeId = employeeId
        self.name = name
        self.role = role
        self.hiredOn = hiredOn
        self.salary = salary
        self.currency = currency
    }
}

/// A leave request.
public struct LeaveRequest: Sendable, Equatable, Codable {
    public let requestId: String
    public let employeeId: String
    public let kind: String
    public let from: Date
    public let to: Date
    public let status: String

    public init(requestId: String, employeeId: String, kind: String, from: Date, to: Date, status: String) {
        self.requestId = requestId
        self.employeeId = employeeId
        self.kind = kind
        self.from = from
        self.to = to
        self.status = status
    }
}

/// A performance review.
public struct PerformanceReview: Sendable, Equatable, Codable {
    public let reviewId: String
    public let employeeId: String
    public let reviewedOn: Date
    public let ratingOutOf5: Int
    public let notes: String

    public init(reviewId: String, employeeId: String, reviewedOn: Date, ratingOutOf5: Int, notes: String) {
        self.reviewId = reviewId
        self.employeeId = employeeId
        self.reviewedOn = reviewedOn
        self.ratingOutOf5 = ratingOutOf5
        self.notes = notes
    }
}

// MARK: - Errors

public enum HRError: Error, Equatable, CustomStringConvertible {
    case unknownLeaveRequest(String)

    public var description: String {
        switch self {
        case .unknownLeaveRequest(let id): return "Unknown leave request \(id)"
        }
    }
}

// MARK: - Contract

/// Hiring, leave management, and performance reviews for the HR vertical.
public protocol IHRBoard: AnyObject, Sendable {
    func hire(_ e: Employee)
    func getEmployee(_ id: String) -> Employee?
    var employees: [Employee] { get }
    func request(_ r: LeaveRequest)
    func decideLeave(requestId: String, decision: String) throws
    func pendingLeaves() -> [LeaveRequest]
    func review(_ r: PerformanceReview)
    func avgRatingFor(_ employeeId: String) -> Double
}

// MARK: - InMemoryHRBoard

/// Deterministic in-memory `IHRBoard`. All state guarded by a single `NSLock`.
public final class InMemoryHRBoard: IHRBoard, @unchecked Sendable {
    private let lock = NSLock()
    private var employeesMap: [String: Employee] = [:]
    private var leaves: [String: LeaveRequest] = [:]
    private var reviews: [PerformanceReview] = []

    public init() {}

    public func hire(_ e: Employee) {
        lock.lock(); defer { lock.unlock() }
        employeesMap[e.employeeId] = e
    }

    public func getEmployee(_ id: String) -> Employee? {
        lock.lock(); defer { lock.unlock() }
        return employeesMap[id]
    }

    public var employees: [Employee] {
        lock.lock(); defer { lock.unlock() }
        return employeesMap.values.sorted { $0.name < $1.name }
    }

    public func request(_ r: LeaveRequest) {
        lock.lock(); defer { lock.unlock() }
        leaves[r.requestId] = r
    }

    public func decideLeave(requestId: String, decision: String) throws {
        lock.lock(); defer { lock.unlock() }
        guard let r = leaves[requestId] else { throw HRError.unknownLeaveRequest(requestId) }
        leaves[requestId] = LeaveRequest(requestId: r.requestId, employeeId: r.employeeId, kind: r.kind,
                                         from: r.from, to: r.to, status: decision)
    }

    public func pendingLeaves() -> [LeaveRequest] {
        lock.lock(); defer { lock.unlock() }
        return leaves.values.filter { $0.status.caseInsensitiveCompare("Pending") == .orderedSame }
    }

    public func review(_ r: PerformanceReview) {
        lock.lock(); defer { lock.unlock() }
        reviews.append(r)
    }

    public func avgRatingFor(_ employeeId: String) -> Double {
        lock.lock(); defer { lock.unlock() }
        let ratings = reviews.filter { $0.employeeId == employeeId }.map { Double($0.ratingOutOf5) }
        return ratings.isEmpty ? 0.0 : ratings.reduce(0.0, +) / Double(ratings.count)
    }
}

// MARK: - HRDomainContext

/// Static domain-context constants for the HR vertical.
public enum HRDomainContext {
    public static let systemPromptSnippet = "[DOMAIN: HR] You are a human resources expert. Help with job description drafting, interview frameworks, performance review templates, disciplinary procedures, leave management, and people analytics. Apply South African labour law principles. Compliance: Labour Relations Act 66/1995, BCEA, EEA, Skills Development Act, POPIA."
    public static let complianceFlags: [String] = ["LRA_66_1995", "BCEA", "EEA", "Skills_Development_Act", "POPIA"]
    public static let suggestedTools: [String] = ["hris", "document_editor", "analytics", "job_boards"]
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/hr/Hr.kt
// Hr.kt
//
// Kotlin port of CircleAI.HR (HRPrimitives.cs + HRDomainContext.cs +
// HRCompanionAdapter.cs) — the C# reference is the EXACT spec. A
// deterministic in-memory HR board: employees, leave requests, and
// performance reviews, plus the domain context + companion adapter.
//
// Fidelity notes:
//   * C# `record` -> Kotlin `data class`; C# `DateTime`/`DateTimeOffset` -> `Instant`.
//   * C# `decimal` -> `java.math.BigDecimal`.
//   * `Employees` orders by Name ASC; `PendingLeaves` filters Status == "Pending"
//     (OrdinalIgnoreCase); `DecideLeave` throws on unknown request.
//   * `AvgRatingFor` averages ratings for the employee, 0 when none (parity with
//     `DefaultIfEmpty(0).Average()`).

package com.bhengubv.circleai.hr

import com.bhengubv.circleai.companion.CompanionContext
import com.bhengubv.circleai.companion.CompanionProactiveEvent
import com.bhengubv.circleai.companion.CompanionTurn
import com.bhengubv.circleai.companion.ICompanionSession
import com.bhengubv.circleai.companion.InterfaceKind
import kotlinx.coroutines.flow.Flow
import java.math.BigDecimal
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

// =====================================================================
// Primitives (HRPrimitives.cs)
// =====================================================================

/** An employee. Mirrors C# `Employee`. */
data class Employee(
    val employeeId: String,
    val name: String,
    val role: String,
    val hiredOn: Instant,
    val salary: BigDecimal,
    val currency: String,
)

/** A leave request. Mirrors C# `LeaveRequest`. */
data class LeaveRequest(
    val requestId: String,
    val employeeId: String,
    val kind: String,
    val from: Instant,
    val to: Instant,
    val status: String,
)

/** A performance review. Mirrors C# `PerformanceReview`. */
data class PerformanceReview(
    val reviewId: String,
    val employeeId: String,
    val reviewedOn: Instant,
    val ratingOutOf5: Int,
    val notes: String,
)

/** Deterministic HR board. Mirrors C# `IHRBoard`. */
interface IHRBoard {
    fun hire(e: Employee)
    fun getEmployee(id: String): Employee?
    val employees: List<Employee>
    fun request(r: LeaveRequest)
    fun decideLeave(requestId: String, decision: String)
    fun pendingLeaves(): List<LeaveRequest>
    fun review(r: PerformanceReview)
    fun avgRatingFor(employeeId: String): Double
}

/** In-memory [IHRBoard]. Mirrors C# `InMemoryHRBoard`. */
class InMemoryHRBoard : IHRBoard {
    private val employees_ = ConcurrentHashMap<String, Employee>()
    private val leaves = ConcurrentHashMap<String, LeaveRequest>()
    private val reviews = mutableListOf<PerformanceReview>()
    private val lock = Any()

    override fun hire(e: Employee) { employees_[e.employeeId] = e }
    override fun getEmployee(id: String): Employee? = employees_[id]
    override val employees: List<Employee>
        get() = employees_.values.sortedBy { it.name }

    override fun request(r: LeaveRequest) { leaves[r.requestId] = r }

    override fun decideLeave(requestId: String, decision: String) {
        val r = leaves[requestId] ?: throw IllegalStateException("Unknown leave request $requestId")
        leaves[requestId] = r.copy(status = decision)
    }

    override fun pendingLeaves(): List<LeaveRequest> =
        leaves.values.filter { it.status.equals("Pending", ignoreCase = true) }

    override fun review(r: PerformanceReview) { synchronized(lock) { reviews.add(r) } }

    override fun avgRatingFor(employeeId: String): Double = synchronized(lock) {
        val ratings = reviews.filter { it.employeeId == employeeId }.map { it.ratingOutOf5.toDouble() }
        if (ratings.isEmpty()) 0.0 else ratings.average()
    }
}

// =====================================================================
// DomainContext (HRDomainContext.cs)
// =====================================================================

/** Static domain context for HR. Mirrors C# `HRDomainContext`. */
object HRDomainContext {
    const val SYSTEM_PROMPT_SNIPPET: String =
        "[DOMAIN: HR] You are a human resources expert. Help with job description drafting, interview " +
            "frameworks, performance review templates, disciplinary procedures, leave management, and people " +
            "analytics. Apply South African labour law principles. Compliance: Labour Relations Act 66/1995, " +
            "BCEA, EEA, Skills Development Act, POPIA."

    val complianceFlags: List<String> = listOf("LRA_66_1995", "BCEA", "EEA", "Skills_Development_Act", "POPIA")

    val suggestedTools: List<String> = listOf("hris", "document_editor", "analytics", "job_boards")
}

// =====================================================================
// CompanionAdapter (HRCompanionAdapter.cs)
// =====================================================================

/** Wraps an [ICompanionSession] with the HR snippet + helpers. Mirrors C# `HRCompanionAdapter`. */
class HRCompanionAdapter(private val inner: ICompanionSession) : ICompanionSession {
    override val sessionId: String get() = inner.sessionId
    override val identityId: String get() = inner.identityId
    override val interfaceKind: InterfaceKind get() = inner.interfaceKind
    override val history: List<CompanionTurn> get() = inner.history
    override val proactiveEvents: Flow<CompanionProactiveEvent> get() = inner.proactiveEvents

    override fun getContext(): CompanionContext = inner.getContext()
    override suspend fun refreshContextAsync() = inner.refreshContextAsync()
    override suspend fun signalFeedbackAsync(positive: Boolean, note: String?) =
        inner.signalFeedbackAsync(positive, note)
    override fun close() = inner.close()

    override suspend fun sendAsync(message: String): String = inner.sendAsync(enrich(message))
    override fun streamAsync(message: String): Flow<String> = inner.streamAsync(enrich(message))
    override suspend fun agentAsync(instruction: String): String = inner.agentAsync(enrich(instruction))

    private fun enrich(m: String): String = "${HRDomainContext.SYSTEM_PROMPT_SNIPPET}\n\n$m"

    suspend fun draftJobDescriptionAsync(role: String, requirements: String): String =
        inner.agentAsync("Draft a compelling, legally compliant job description for: $role. Requirements: $requirements. Include purpose, responsibilities, qualifications, and EEA statement.")

    suspend fun generatePerformanceReviewAsync(employeeName: String, role: String, achievements: String): String =
        inner.agentAsync("Generate a structured performance review for $employeeName ($role). Achievements: $achievements. Include ratings, development areas, and SMART goals.")

    suspend fun adviseOnDisciplinaryAsync(misconduct: String, employeeHistory: String): String =
        inner.agentAsync("Advise on disciplinary action for: $misconduct. Employee history: $employeeHistory. Apply LRA progressive discipline principles and recommend appropriate sanction.")

    suspend fun draftJobDescriptionAsync(roleTitle: String, seniority: String, mustHaves: String): String =
        inner.agentAsync("Draft a job description for $seniority $roleTitle. Must-haves: $mustHaves. Inclusive language, outcomes-led not task-list.")

    suspend fun structureInterviewLoopAsync(role: String, hoursAvailable: Int): String =
        inner.agentAsync("Structure an interview loop for $role in $hoursAvailable hours. Map each stage to a competency, name the evaluator role.")

    suspend fun writePerformanceFeedbackAsync(employeeName: String, strengths: String, growthAreas: String): String =
        inner.agentAsync("Write performance feedback for $employeeName. Strengths: $strengths. Growth: $growthAreas. SBI format, specific, future-focused.")

    suspend fun handleSensitiveHrIssueAsync(situation: String, jurisdiction: String): String =
        inner.agentAsync("Suggest first-response plan for HR situation: $situation in $jurisdiction. Cover legal hold, witness, documentation, escalation path.")
}
